File: vol2/ch20/LeetCode1102-1.cpp
struct QNode {                  //队列元素类型
    int x,y;                    //位置
    int length;                 //路径长度
};
class Solution {
    int dx[4]={0,0,1,-1};       //水平方向偏移量
    int dy[4]={1,-1,0,0};       //垂直方向偏移量
public:
    int maximumMinimumPath(vector<vector<int>>& grid) {
        int m=grid.size(),n=grid[0].size();
        vector<vector<int>> dist(m,vector<int>(n,-1));
        queue<QNode> qu;
        QNode e,e1;
        e.x=0;e.y=0;e.length=grid[0][0];
        qu.push(e);
        dist[0][0]=grid[0][0];
        int ans=0;
        while(!qu.empty()) {
            e=qu.front();qu.pop();      //出队元素e
            int x=e.x,y=e.y,length=e.length;
            if(x==m-1 && y==n-1) ans=max(ans,length);
            for(int di=0;di<4;di++) {
                int nx=x+dx[di];
                int ny=y+dy[di];
                if(nx<0 || nx>=m || ny<0 || ny>=n) continue;
                int curlength=min(dist[x][y],grid[nx][ny]);
                if(curlength>dist[nx][ny]) {    //剪支1
                    dist[nx][ny]=curlength;
                    if(dist[nx][ny]>ans) {      //剪支2
                        e1.x=nx;e1.y=ny;e1.length=dist[nx][ny];
                        qu.push(e1);
                    }
                }
            }
        }
        return ans;
    }
};

File: vol2/ch20/LeetCode1102-1.py
class Solution:
    def maximumMinimumPath(self, grid: List[List[int]]) -> int:
        dx=[0,0,1,-1]                       #水平方向偏移量
        dy=[1,-1,0,0]                   	#垂直方向偏移量
        m,n=len(grid),len(grid[0])
        dist=[[-1 for _ in range(0,n)] for _ in range(0,m)]
        qu=deque()               #定义一个队列,元素为[x,y,length]
        qu.append([0,0,grid[0][0]])      #源点元素进队
        dist[0][0]=grid[0][0]
        ans=0
        while qu:
            [x,y,length]=qu.popleft()       #出队元素
            if x==m-1 and y==n-1:ans=max(ans,length)
            for di in range(0,4):
                nx,ny=x+dx[di],y+dy[di]
                if nx<0 or nx>=m or ny<0 or ny>=n:continue
                curlength=min(dist[x][y],grid[nx][ny])
                if curlength>dist[nx][ny]:			#剪支1
                    dist[nx][ny]=curlength
                    if dist[nx][ny]>ans:            #剪支2
                        qu.append([nx,ny,dist[nx][ny]])
        return ans
